core/src/test/java/com/linecorp/armeria/common/stream/ByteStreamMessageOutputStreamTest.java

```diff
@@ -0,0 +1,317 @@
+                    .thenAwait(Duration.ofSeconds(3))
+                    .then(() -> await().untilAtomic(count, Matchers.is(2)))
+                    .thenRequest(1)
+                    .expectNext(httpData(2))
+                    .verifyComplete();
+    }
+
+    @Test
+    void writeAfterClosed() {
+        final ByteStreamMessage byteStreamMessage = StreamMessage.fromOutputStream(os -> {
+            try {
+                for (int i = 0; i < 5; i++) {
+                    os.write(i);
+                }
+                os.close();
+
+                assertThatThrownBy(() -> os.write(0))
+                        .isInstanceOf(IOException.class)
+                        .hasMessage("Stream closed");
+                assertThatThrownBy(() -> os.write(new byte[] { 0 }, 0, 1))
+                        .isInstanceOf(IOException.class)
+                        .hasMessage("Stream closed");
+                assertThatCode(os::close).doesNotThrowAnyException();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        });
+
+        StepVerifier.create(byteStreamMessage)
+                    .expectNext(httpData(0), httpData(1), httpData(2), httpData(3), httpData(4))
+                    .verifyComplete();
+    }
+
+    @Test
+    void writeAfterStreamClosed() throws InterruptedException {
+        final CountDownLatch wait = new CountDownLatch(1);
+        final CountDownLatch end = new CountDownLatch(1);
+        final ByteStreamMessage byteStreamMessage = StreamMessage.fromOutputStream(os -> {
+            try {
+                for (int i = 0; i < 5; i++) {
+                    if (i < 2) {
+                        os.write(i);
+                    } else {
+                        wait.await();
+                        assertThatThrownBy(() -> os.write(0))
+                                .isInstanceOf(IOException.class)
+                                .hasMessage("Stream closed");
+                    }
+                }
+                assertThatCode(os::close).doesNotThrowAnyException();
+                end.countDown();
+            } catch (IOException | InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        });
+
+        StepVerifier.create(byteStreamMessage, 2)
+                    .expectNext(httpData(0), httpData(1))
+                    .then(byteStreamMessage::abort)
+                    .then(wait::countDown)
+                    .verifyError(AbortedStreamException.class);
+        end.await();
+    }
+
+    @Test
+    void writeAfterAborted() {
+        final StreamMessage<Integer> streamMessage = StreamMessage.of(10, 11, 12, 13, 14);
+        final StreamMessage<HttpData> aborted = streamMessage
+                .peek(x -> {
+                    if (x == 13) {
+                        streamMessage.abort();
+                        await().untilAsserted(() -> assertThat(streamMessage.isOpen()).isFalse());
```

Comment:
A `FixedStreamMessage` is always closed because all elements are given when it is created.
https://github.com/line/armeria/blob/871d87297e4d051241589cb1ae95641cbc83f880/core/src/main/java/com/linecorp/armeria/internal/common/stream/FixedStreamMessage.java#L87

If you want to know whether all elements of the stream are drained by `abort()`, you can use `.whenComplete().join()` and check the raised exception.

Reply:
✅ updated, https://github.com/line/armeria/pull/4402/commits/7f284df053d8c3f3dfcac7834478ad9d35258a9e